With lazy loading of contributions

maskaravivek opened the pull request:
Fixes #2904
Fixes #1993
Fixes #3667
Fixes #52

macgills commented on app/src/main/java/fr/free/nrw/commons/di/CommonsApplicationModule.java:
```diff
@@ -231,10 +234,55 @@ public QuadTree providesQuadTres() {
+    static final Migration MIGRATION_1_2 = new Migration(1, 2) {
+        @Override
+        public void migrate(SupportSQLiteDatabase database) {
+            database.execSQL("DROP TABLE `contribution`");
```
If we just delete everything instead of migrating data we could `fallbackDestructively` like mentioned [here](https://developer.android.com/training/data-storage/room/migrating-db-versions#handle-missing-migrations)

maskaravivek replied:
Thanks, @macgills. Will make this change.

macgills commented on .idea/codeStyles/codeStyleConfig.xml:
```diff
@@ -1,5 +1,6 @@
 <component name="ProjectCodeStyleConfiguration">
   <state>
     <option name="USE_PER_PROJECT_SETTINGS" value="true" />
+    <option name="PREFERRED_PROJECT_CODE_STYLE" value="Default" />
```
This might mean you are not using the Project Default?

Please verify your settings are like this 
![image](https://user-images.githubusercontent.com/3358282/79971934-06f20180-848d-11ea-99d0-90ec5c89d504.png)

maskaravivek replied:
I have been using project default for inspection. Also, these inspections are working for me as studio now points out several inspection issues which it didn't do earlier. 

<img width="1018" alt="Screen Shot 2020-04-22 at 9 45 04 PM" src="https://user-images.githubusercontent.com/3069373/80060100-8bbb3a80-84e2-11ea-892d-627dc605f73c.png">

macgills replied:
My apologies, I screenshot the wrong thing

![image](https://user-images.githubusercontent.com/3358282/80078085-50eaee00-8546-11ea-8040-a47883601685.png)

maskaravivek replied:
That too is set to use project settings. 

<img width="1016" alt="Screen Shot 2020-04-23 at 1 44 21 AM" src="https://user-images.githubusercontent.com/3069373/80078605-fe88dd80-8503-11ea-8195-714d1f0ffd0d.png">

Everytime i switch branch I keep stashing this change. Is there a way to fix it?

macgills replied:
![image](https://user-images.githubusercontent.com/3358282/80079906-ca83db80-8548-11ea-8076-2e7e5133f93b.png)

This is the change that happens when I set my code style to `default`

If you just discard this change does it come back? You can try switching codestyles a bit (and clicking apply) and see what happens

macgills replied:
I found a potential [solution](https://youtrack.jetbrains.com/issue/IDEA-207303#focus=streamItem-27-3299338.0-0) that may work?

macgills commented on app/src/main/java/fr/free/nrw/commons/Media.java:
```diff
@@ -50,6 +51,8 @@
      */
+    @PrimaryKey
+    @NonNull
     private String pageId;
```
I think pageId is frequently null? I had that Pr where we kept requesting `Mnull`

maskaravivek replied:
It won't be null for Media. It can be null for other API calls where `MwQueryPage` is returned as response.

macgills replied:
Then `NonNull` is not very true if it can be null. Is the annotation a Room requirement on primary keys? I've forgotten

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionsListFragment.java:
```diff
@@ -99,14 +105,21 @@ public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
+        LinearLayoutManager layoutManager;
         if (getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE) {
-            rvContributionsList.setLayoutManager(new GridLayoutManager(getContext(),SPAN_COUNT));
+            layoutManager = new GridLayoutManager(getContext(), SPAN_COUNT);
```
a GridLayoutManager with spanCount 1 is essentially analogous to LinearLayoutManager, I would believe

That way we could move this condition into a ternary and not create different types of LayoutManager

maskaravivek replied:
We are using span count = 3.

macgills replied:
Yes we are but Instead of a LinearLayoutManager and a GridLayoutManager we could use only a GridLayoutManager and based on Orientation set the span count

macgills commented on app/src/main/java/fr/free/nrw/commons/media/MediaInterface.java:
```diff
@@ -17,6 +15,11 @@
             "&iiextmetadatafilter=DateTime|Categories|GPSLatitude|GPSLongitude|ImageDescription|DateTimeOriginal" +
             "|Artist|LicenseShortName|LicenseUrl";
+
+    String USER_MEDIA_PARAMS = "&prop=imageinfo&iiprop=url|extmetadata&iiurlwidth=640" +
```
is thisdifferent from the above params?

macgills commented on app/src/main/java/fr/free/nrw/commons/media/MediaInterface.java:
```diff
@@ -48,6 +51,21 @@
+    /**
+     * This method retrieves a list of Media objects for a given user name
+     *
+     * @param username     the category name. Must start with "Category:"
```
the username is... the category name?

macgills commented on app/src/main/java/fr/free/nrw/commons/upload/UploadService.java:
```diff
@@ -162,7 +162,7 @@ public void queue(int what, Contribution contribution) {
                         .observeOn(mainThreadScheduler)
                         .subscribe(aLong->{
-                            contribution.set_id(aLong);
+                            contribution.setPageId(String.valueOf(aLong));
```
Not sure about this at all, at this point a contribution can't have a pageID as it doesn't exist on the servers, we try and save it with nothing(?) set as pageId? I am pretty sure `@AutoGenerate` has been removed so that won't work

misaochan commented:
Just to clarify, we are waiting on implementation of the paging library #3468 for this, right? Or shall we take this as it is and modify later?

maskaravivek commented:
@misaochan Contributions doesn't use `RvRenderer` currently. It simply uses Recycler View's View holder which i assume should be compatible with the paging library.

maskaravivek commented:
@misaochan @neslihanturan This PR is now ready for review.

neslihanturan commented:
Hey @maskaravivek , why Travis fails for this one?

maskaravivek commented:
@neslihanturan I have just fixed the build.

ashishkumar468 commented on app/src/main/java/fr/free/nrw/commons/OkHttpConnectionFactory.java:
```diff
@@ -45,7 +46,7 @@ private static OkHttpClient createClient() {
     private static HttpLoggingInterceptor getLoggingInterceptor() {
         final HttpLoggingInterceptor httpLoggingInterceptor = new HttpLoggingInterceptor()
-                .setLevel(Level.BASIC);
+            .setLevel(Level.BODY);
```
Is this intentional?

ashishkumar468 commented on app/src/main/java/fr/free/nrw/commons/contributions/Contribution.java:
```diff
@@ -161,17 +143,26 @@ public int describeContents() {
+            media.getDataLength(), media.getDateCreated(), media.getDateUploaded(),
+            media.getLicense(), media.getLicenseUrl(), media.getCreator(), media.getCategories(),
+            media.isRequestedDeletion(), media.getCoordinates());
+        this.state = STATE_COMPLETED;
```
Maybe add a comment, why this constructor has state_completed by default

macgills commented on app/src/main/java/fr/free/nrw/commons/Media.java:
```diff
@@ -64,14 +68,28 @@
      */
-    public Media(String filename) {
+    public Media(final String filename) {
+        pageId = UUID.randomUUID().toString();
```
you could replace this line with `this()`

maskaravivek replied:
@macgills I didn't understand the requested change. Can you elaborate?

macgills replied:
`pageId = UUID.randomUUID().toString();` is repeated, you can invoke the parameter-less constructor by calling `this()` and stop the repetition.

Kotlin - because it is great - enforces this and all secondary constructors must delegate to the primary constructor.

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionsFragment.java:
```diff
@@ -307,43 +226,47 @@ void onAuthCookieAcquired() {
     @Override
-    public int getTotalMediaCount() {
-        return numberOfContributions;
+    public void refreshSource() {
```
this method can probably be deleted?

macgills commented on app/src/main/java/fr/free/nrw/commons/upload/UploadService.java:
```diff
@@ -299,11 +297,12 @@ private void saveCompletedContribution(Contribution contribution, UploadResult u
-            .observeOn(mainThreadScheduler)
-            .subscribe());
+
+        final Disposable disposable = contributionDao.delete(contribution)
```
do we need this variable? It could be inlined like the previous code

misaochan commented:
Will test this once the code review issues have been addressed.

macgills commented:
Should this be using a [BoundaryCallback](https://developer.android.com/reference/androidx/paging/PagedList.BoundaryCallback) ?

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionsListAdapter.java:
```diff
@@ -1,52 +1,54 @@
+
+    private static DiffUtil.ItemCallback<Contribution> DIFF_CALLBACK =
+        new DiffUtil.ItemCallback<Contribution>() {
+            // Concert details may have changed if reloaded from the database,
```
"concert" in variable names and comments

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionsListAdapter.java:
```diff
@@ -1,52 +1,54 @@
+            @Override
+            public boolean areContentsTheSame(Contribution oldConcert,
+                Contribution newConcert) {
+                return oldConcert.equals(newConcert);
```
So you overrode equals to get the effect of `old.pageId==new.pageId` here by converting them to strings first. Couple problems
1. Quite surprising, could have effects at any other comparison site
1. Android studio will create hashCode, equals & toString methods for you, press `alt+insert`
1. no hashcode implementation was added

To make explicit your intention you could have said
```
            @Override
            public boolean areContentsTheSame(Contribution oldConcert,
                Contribution newConcert) {
                return areItemsTheSame(oldConcert, newConcert);
```

which is I think what the intention was.

This leads in to the larger problem of a contribution can't change its contents which they should be able to do I think? This could lead to some very confusing errors where the data has been updated and it just won't render

maskaravivek replied:
Agreed. I didn't notice this while making the changes. I have updated the `hashcode` and `equals` implementation for `Media` and `Contribution` classes. It would be great if you could take a look.

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionsListFragment.java:
```diff
@@ -211,23 +233,25 @@ public void onSaveInstanceState(@NonNull Bundle outState) {
     public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
         super.onViewStateRestored(savedInstanceState);
         if(null!=savedInstanceState){
-            lastVisibleItemID =savedInstanceState.getString(VISIBLE_ITEM_ID, null);
```
not restoring the state?

maskaravivek replied:
The scrolled state of the list is still being retained. I am not sure how this variable was being used earlier.

macgills commented on app/src/main/java/fr/free/nrw/commons/media/MediaClient.java:
```diff
@@ -83,6 +84,24 @@ public MediaClient(MediaInterface mediaInterface, MediaDetailInterface mediaDeta
+     */
+    public Single<List<Media>> getMediaListForUser(String userName) {
+        return responseToMediaList(
+            continuationStore.containsKey("user_" + userName) ?
```
this is a bit hard on the eyes, how about moving the ternary into the method call?

```
mediaInterface.getMediaListForUser(
username,
10,
continuationStore.hasUser ? continuation.getUser : emptyMap()
)
```

macgills commented on app/src/test/kotlin/fr/free/nrw/commons/contributions/ContributionsPresenterTest.kt:
```diff
@@ -60,17 +60,6 @@ class ContributionsPresenterTest {
-     * Test fetch contributions
-     */
-    @Test
-    fun testFetchContributions(){
```
This functionality hasn't been deleted just moved, right? It would be a shame to lose the coverage

maskaravivek replied:
The functionality has changed but I have added a new test case for this method. I will add more test cases once the functionality gets tested.

maskaravivek commented:
Thanks for the review @macgills. I have addressed the comments. 

It would be great if a couple of people could help in testing the PR's functionality. Am not sure if everything is working perfectly. It would be nice to have bugs pointed out. 

I will take up java docs, test cases, etc once the PR looks good from the functionality point of view.

maskaravivek commented:
> Should this be using a [BoundaryCallback](https://developer.android.com/reference/androidx/paging/PagedList.BoundaryCallback) ?

Yes, it should. Making this change.

maskaravivek commented:
Converting the PR to a draft PR. Things are not working after adding the paging library. I will fix it and notify you here when it is ready for testing.

misaochan commented:
> It would be great if a couple of people could help in testing the PR's functionality. Am not sure if everything is working perfectly. It would be nice to have bugs pointed out.

Sure, happy to do that, ping me when you have fixed the paging library. :)

maskaravivek commented:
The PR is now ready for review.

maskaravivek commented on app/build.gradle:
```diff
@@ -121,6 +125,8 @@ dependencies {
     // Kotlin
     implementation "androidx.preference:preference-ktx:$PREFERENCE_VERSION"
+
+    implementation 'androidx.multidex:multidex:2.0.1'
```
I had to enable multidex support as the number of methods grew after adding a couple of dependencies.

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionsFragment.java:
```diff
@@ -386,10 +297,6 @@ public void onResume() {
 
         boolean isSettingsChanged = store.getBoolean(Prefs.IS_CONTRIBUTION_COUNT_CHANGED, false);
```
this could be deleted now, right?

macgills replied:
IS_CONTRIBUTION_COUNT_CHANGED is now only used to put the boolean, never retrieve it, can be deleted

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionsListAdapter.java:
```diff
@@ -1,68 +1,74 @@
         this.callback = callback;
         this.mediaClient = mediaClient;
-        contributions = new ArrayList<>();
+        setHasStableIds(false);
```
this is the default value so could be omitted

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionsListFragment.java:
```diff
@@ -16,218 +17,218 @@
+public class ContributionsListFragment extends CommonsDaggerSupportFragment implements
+    ContributionsListContract.View, ContributionsListAdapter.Callback {
+
+  private static final String VISIBLE_ITEM_ID = "visible_item_id";
```
unused I think

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionsListPresenter.java:
```diff
@@ -0,0 +1,76 @@
+  void setup() {
+    final PagedList.Config pagedListConfig =
+        (new PagedList.Config.Builder())
+            .setEnablePlaceholders(true)
```
I think we shouldn't use placeholders, it sounds like we would want to know the entire data set beforehand? [docs](https://developer.android.com/topic/libraries/architecture/paging/ui#provide-placeholders)

macgills commented on app/src/test/kotlin/fr/free/nrw/commons/contributions/ContributionsRepositoryTest.kt:
```diff
@@ -0,0 +1,55 @@
+        val contributions = listOf(mock(Contribution::class.java))
+        whenever(localDataSource.saveContributions(ArgumentMatchers.anyList()))
+            .thenReturn(Single.just(listOf(1L)))
+        val save = contributionsRepository.save(contributions).blockingGet()
```
you can use `.test().assert*` instead

macgills replied:
@maskaravivek this change could also be made

macgills commented on app/src/test/kotlin/fr/free/nrw/commons/media/MediaClientTest.kt:
```diff
@@ -197,6 +197,35 @@ class MediaClientTest {
+        val imageInfo = ImageInfo()
+
+        val mwQueryPage = mock(MwQueryPage::class.java)
+        `when`(mwQueryPage.title()).thenReturn("Test")
```
`whenever`

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionBoundaryCallback.java:
```diff
@@ -0,0 +1,112 @@
+  public void fetchContributions() {
+    networkState.postValue(NetworkState.LOADING);
+    compositeDisposable.add(mediaClient.getMediaListForUser(sessionManager.getUserName())
+        .map(mediaList -> {
```
this is a whole new class, it would be much nicer in kotlin to just use `map` instead of this for loop

macgills replied:
@maskaravivek any inclination to convert this to kotlin?

macgills commented on app/src/test/kotlin/fr/free/nrw/commons/contributions/ContributionBoundaryCallbackTest.kt:
```diff
@@ -0,0 +1,101 @@
+            Single.just(listOf(mock(Media::class.java)))
+        )
+        contributionBoundaryCallback.onZeroItemsLoaded()
+        verify(repository, times(1)).save(anyList());
```
`verify(_, times(1)` is equivalent to `verify(_)`

macgills replied:
`anyList()` is hiding the logic of mapping a media to a completed Contribution

macgills replied:
There is no testing the error case sending a network state or any test involving network state

macgills replied:
>anyList() is hiding the logic of mapping a media to a completed Contribution

This issue is still relevant @maskaravivek

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionBoundaryCallback.java:
```diff
@@ -0,0 +1,112 @@
+   * Saves the contributions the the local DB
+   */
+  private void saveContributionsToDB(final List<Contribution> contributions) {
+    Single<List<Long>> single = repository.save(contributions);
```
this does nothing without a subscribe

macgills replied:
This still does nothing without a subscribe

maskaravivek replied:
The contribution is getting saved in the DB. I didn't understand the comment.

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionBoundaryCallback.java:
```diff
@@ -0,0 +1,112 @@
+      final SessionManager sessionManager,
+      final MediaClient mediaClient,
+      @Named(CommonsApplicationModule.IO_THREAD) final Scheduler ioThreadScheduler) {
+    super();
```
unnecessary super

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionBoundaryCallback.java:
```diff
@@ -0,0 +1,112 @@
+  private final ContributionsRepository repository;
+  private final SessionManager sessionManager;
+  private final MediaClient mediaClient;
+  private final MutableLiveData networkState;
```
raw type of mutable live data

macgills commented on app/src/test/kotlin/fr/free/nrw/commons/media/MediaClientTest.kt:
```diff
@@ -197,6 +197,35 @@ class MediaClientTest {
 
+    @Test
+    fun getMediaListForUser() {
+        val mockContinuation= mapOf(Pair("gcmcontinue", "test"))
```
mapOf("key" to "value") is a bit more idiomatic kotlin

macgills commented on app/src/test/kotlin/fr/free/nrw/commons/media/MediaClientTest.kt:
```diff
@@ -197,6 +197,35 @@ class MediaClientTest {
+        assertEquals(continuationCaptor.allValues[0], emptyMap<String, String>())
+        assertEquals(continuationCaptor.allValues[1], mockContinuation)
+
+        assertEquals(media1.filename, "Test")
```
the mocked object already exists, why are we fetching the title to assert it is the same object?

macgills commented on app/src/test/kotlin/fr/free/nrw/commons/media/MediaClientTest.kt:
```diff
@@ -197,6 +197,35 @@ class MediaClientTest {
+        val media1 = mediaClient!!.getMediaListForUser("Test").blockingGet().get(0)
+        val media2 = mediaClient!!.getMediaListForUser("Test").blockingGet().get(0)
+
+        assertEquals(continuationCaptor.allValues[0], emptyMap<String, String>())
```
you could do this with a verification call instead and it would probably read cleaner

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionBoundaryCallback.java:
```diff
@@ -0,0 +1,96 @@
+   * Saves the contributions the the local DB
+   */
+  private void saveContributionsToDB(final List<Contribution> contributions) {
+    repository.save(contributions);
```
I'll move this discussion to the more recent code.
This does nothing without a subscribe, this is just a Single which represents a task that produces a value or an error.
If the contributions are being saved then it is not by this code

maskaravivek replied:
@macgills There's just a single usage of this method in our codebase. I checked `ContributionsRepository`, `ContributionsLocalDataSource` and `ContributionDao` too. 

Also, I have been testing on a fresh install. 

Can you demonstrate that a `Single` doesn't execute if the result isn't subscribed?

maskaravivek replied:
Anyways, I have updated the code to add a subscription. Still, it would be great if you could cite a source or demonstrate the above.

macgills replied:
This is just an object
![image](https://user-images.githubusercontent.com/3358282/81782750-2f03dc00-94f2-11ea-827b-00c427348723.png)

This is the single executing

![image](https://user-images.githubusercontent.com/3358282/81782838-5195f500-94f2-11ea-9289-1cdb7d9e302b.png)

Or we could check the documentation of [Single](http://reactivex.io/documentation/single.html). This is the behaviour of all the reactive types, you need a subscribe call to actually do/receive something. 

[Tutorial](https://medium.com/@factoryhr/understanding-java-rxjava-for-beginners-5eacb8de12ca) [official examples](https://github.com/ReactiveX/RxJava/wiki/How-To-Use-RxJava)

maskaravivek replied:
Thanks for the example. Hopefully, the updated code should work as expected.

maskaravivek commented:
@macgills I have addressed your other comments too. It would be great if you could take a look at the PR again. Apologies for making you review it multiple times.

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionsFragment.java:
```diff
@@ -384,12 +295,7 @@ public void onResume() {
         locationManager.addLocationListener(this);
 
-        boolean isSettingsChanged = store.getBoolean(Prefs.IS_CONTRIBUTION_COUNT_CHANGED, false);
         store.putBoolean(Prefs.IS_CONTRIBUTION_COUNT_CHANGED, false);
```
`Prefs.IS_CONTRIBUTION_COUNT_CHANGED` can be removed entirely, we now only ever set the value, never retrieve

maskaravivek replied:
Yeah, I think the whole of upload limit setting can be removed now.

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionsListAdapter.java:
```diff
@@ -1,68 +1,73 @@
-            callback.fetchMediaUriFor(contribution);
+    public void onBindViewHolder(@NonNull final ContributionViewHolder holder, final int position) {
+        final Contribution contribution = getItem(position);
+        if (contribution != null) {
```
we no longer need this null check as we don't use placeholders

macgills commented on app/src/main/java/fr/free/nrw/commons/media/MediaClient.java:
```diff
@@ -83,6 +83,22 @@ public MediaClient(MediaInterface mediaInterface, MediaDetailInterface mediaDeta
+    public Single<List<Media>> getMediaListForUser(String userName) {
+        Map<String, String> continuation =
+            continuationStore.containsKey("user_" + userName) ? continuationStore
+                .get("user_" + userName) : Collections.emptyMap();
```
this looks a little funky with the `.get` on the next line, sometimes I favor
```
expression
 ? trueValue
 : falseValue
```

when things run a little long

macgills commented on app/src/test/kotlin/fr/free/nrw/commons/media/MediaClientTest.kt:
```diff
@@ -197,6 +203,38 @@ class MediaClientTest {
+        val media1 = mediaClient!!.getMediaListForUser("Test").blockingGet().get(0)
+        val media2 = mediaClient!!.getMediaListForUser("Test").blockingGet().get(0)
+
+        verify(mediaInterface, times(2))?.getMediaListForUser(
```
my intention with the verify was to not use argument matchers and have 2 separate calls that verify exact arguments.

I imagine this test invokes `getMediaListForUser` twice for the purpose of testing the continuationStore ternary? `anyMap()` is hiding this fact.

A more robust - but unnecessary - way to approach this would be if continuationStroe was not a hashMap but an object ContinuationStore we could mock and expect values from and then we would have separate tests for when there is and isn't a match in the continuation store

maskaravivek replied:
Yeah, I agree. This test and other similar tests in `MediaClientTest` can be improved. Can you create a separate issue for it? I have just mimicked the existing `getMediaListFromCategoryTwice` test(i am the original of that test too).

These tests can probably be improved in different PR.

macgills replied:
https://github.com/commons-app/apps-android-commons/issues/3749 done

macgills commented on app/build.gradle:
```diff
@@ -44,6 +44,10 @@ dependencies {
 
+    //paging
+    implementation "androidx.paging:paging-runtime:2.1.2"
+    implementation "androidx.paging:paging-rxjava2:2.1.2"
```
there is a testImplementation of the pagingLibrary as well with no Android dependencies, was that not useful?

maskaravivek replied:
Can you point me to the library? I didn't actually understand this comment.

macgills replied:
The androidx team provide a library without Android Dependencies: [paging-commons](https://developer.android.com/topic/libraries/architecture/paging#setup)

I can't find any example of why it is useful? I would have thought given the `testImplementation` that it is for writing unit tests involving PagedList etc.

Also we should be using the `-ktx` variants of these libraries, it'll give us some advantages when writing kotlin

maskaravivek replied:
I have added the `-ktx` version of the library. For writing paged list tests I have added a few methods in `PagedListMock.kt` class. That seems to be serving the purpose for now. We can add the test implementation later if required.

macgills commented:
> @macgills I have addressed your other comments too. It would be great if you could take a look at the PR again. Apologies for making you review it multiple times.

No worries, apologies for being a stickler.

There are some comments that weren't addressed, very hard to see in such a large PR, I recommend scrolling in the conversation tab - making sure to expand "hidden items" - as I have gone through resolving the appropriate comments and tagging you where I think there may be some actionable thing.

There are also additional comments but nothing major, I'll get to testing soon

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionBoundaryCallback.kt:
```diff
@@ -0,0 +1,90 @@
+    fun fetchContributions() {
+        compositeDisposable.add(
+            mediaClient.getMediaListForUser(sessionManager.userName)
+                .map<List<Contribution>> { mediaList: List<Media?> ->
```
you can remove the type information from this `map`

also my original point here was we could map the mediaList and avoid the overly verbose loop eg
```
mediaList.map { Contribution(it, Contribution.STATE_COMPLETED) }
```

and the OnNext action in the subscribe can become a method reference

```
.subscribe(
   ::saveContributionsToDb, 
```

macgills commented:
When uploading a contribution it shows in the list and then disappears.

maskaravivek commented:
@macgills I have fixed the issue. Apparently I wasn't subscribing to the `Single` that was saving the contribution after upload.

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionBoundaryCallback.kt:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    /**
+     * It is triggered when the user scrolls to the top of the list No action is taken at this point
```
this comment is no longer true

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionsListAdapter.java:
```diff
@@ -1,68 +1,71 @@
-    public void setContributions(@NonNull List<Contribution> contributionList) {
-        contributions = contributionList;
-        notifyDataSetChanged();
+    Contribution getContributionForPosition(final int position) {
```
do we need this method when we have `getItem` ?

maskaravivek replied:
`getItem` is protected. We are using `getContributionForPosition` outside the adapter class.

macgills replied:
yup, my mistake

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/NetworkState.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Enum that defines the different states for contribution paged list's boundary callback
+ */
+enum NetworkState {
```
never used

macgills commented:
I am seeing a few of these logs

```
2020-05-20 13:23:45.171 13978-14916/fr.free.nrw.commons.beta E/ContributionBoundaryCallback$fetchContributions: Failed to fetch contributions: Query map was null (parameter #3)
        for method MediaInterface.getMediaListForUser
```

should we be null checking

```
continuationStore.put(key, mwQueryResponse.continuation());
```

I haven't looked much at the structure of the returned response and what precisely a continuation is

maskaravivek commented:
@macgills Thanks for testing the PR again. I have fixed the issue and have also addressed other code review comments. 

> Query map was null
Earlier, the end of the list was not handled so this error was coming.

macgills commented on app/src/main/java/fr/free/nrw/commons/media/MediaClient.java:
```diff
@@ -123,7 +140,12 @@ public MediaClient(MediaInterface mediaInterface, MediaDetailInterface mediaDeta
             }
-            continuationStore.put(key, mwQueryResponse.continuation());
+            if(mwQueryResponse.continuation() != null) {
+                continuationStore.put(key, mwQueryResponse.continuation());
```
if null couldn't we just remove the key and then it all works as usual without the extra map

maskaravivek replied:
If we remove the key then it would be like fetching the first page. The API call would be made without the continuation params.

macgills replied:
And then the Db update would return the same data, not retriggering a boundaryCallback as the list would be unchanged, I think.

Though I guess now the api is set up with a continuation... I am quite unsure of how to proceed

macgills replied:
and the continuation map is never assigned but I am sure you are on that

maskaravivek replied:
> Though I guess now the api is set up with a continuation... I am quite unsure of how to proceed

Yeah, right. As the API is set up with a continuation it will start fetching all the pages again. On a related note, the current logic of fetching pages(that I added in this PR) is very naive and can be improved. For now, i wanted to keep things simple. 

> and the continuation map is never assigned but I am sure you are on that

I didn't understand this comment.

macgills replied:
I should have said `continuationExists` map but it is resolved now.

This is a bit "upending the teatable" but it seems to me that this endpoint shouldn't use continuation at all, we could happily query using

![image](https://user-images.githubusercontent.com/3358282/82648690-f2ca2d00-9c0f-11ea-8829-e9d92e9d69e0.png)

but perhaps that is best suited for another ticket

maskaravivek replied:
Yeah we can use `aistart` and `aiend` parameter to fetch contributions based on timestamp but I don't think that can replace continuation. With start and end timestamp, there's no way to be sure how many contributions would be returned. For eg. recently I uploaded around 2000 images in a day. The API would then return all images in one call. It would defeat the purpose of pagination. 

To limit it to 10 images per call we will have to use `ailimit` and then `continuation` will still come into picture.

maskaravivek replied:
@macgills Is everything working in this PR now? It would be great if you could do another test.

macgills replied:
It is a little trickier granted but if at top of list we request media since now and reorder it so it is oldest first it hoovers along nicely.

Bottom of the list request since last item timestamp + newest first.

Again perhaps a discussion for another ticket. I will get to testing shortly

macgills commented:
The ux after uploading is not as good, you have to scroll up to see the in progress upload and then when it is done the list jumps down to the next item and you have to scroll up again to see the now uploaded item. Maybe if the deleting and saving of the inProgress/complete Contribution were in a transaction it would look better.

maskaravivek commented:
@macgills I have fixed the UI issues. I have made the following changes: 
- am deleting and saving in a single transaction
- have replaced a couple of usages of `save` with `update`
- started persisting time information in `dateUploaded` column of Contribution DB. Earlier just date was persisted so all contributions uploaded on a particular day had the same timestamp.

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionsListPresenter.java:
```diff
@@ -62,9 +62,7 @@ public void onDetachView() {
-    compositeDisposable.add(repository.deleteContributionFromDB(contribution)
-        .subscribeOn(ioThreadScheduler)
-        .subscribe());
+    ioThreadScheduler.scheduleDirect(() -> repository.deleteContributionFromDB(contribution));
```
`Completeable.fromAction` may be more idiomatic

macgills commented:
The UI is better but we still don't scroll to the new uploading item however all things considered I think that is an acceptable concession.

I am happy with this PR once `ioThreadScheduler.scheduleDirect` is replaced with completeables that are added to a compositeDisposable so they can be gracefully cancelled.

May be for the best to return these from the dao eg

```
@Insert
public void saveSynchronous()
@Delete
public void deleteSynchronous()

public Completeable saveAndDelete(){
return Completeable.fromAction{
deleteSynchronous()
saveSynchronous()
}
}
```

maskaravivek commented:
Thanks for testing the PR @macgills. 

I have updated the PR.

macgills commented on app/src/main/java/fr/free/nrw/commons/upload/UploadService.java:
```diff
@@ -297,7 +298,7 @@ private void onSuccessfulUpload(Contribution contribution, UploadResult uploadRe
         compositeDisposable.add(mediaClient.getMedia("File:" + uploadResult.getFilename())
         .map(media -> new Contribution(media, Contribution.STATE_COMPLETED))
-        .subscribe(newContribution -> contributionDao.deleteAndSaveContribution(contribution, newContribution)));
+        .subscribe(newContribution -> contributionDao.saveAndDelete(contribution, newContribution)));
```
this won't work, `saveAndDelete` is a completable.

`subscribe` is a terminal call, it receives the values from the particular stream it subscribes to but it cannot continue the stream.

you probably want to use `.flatMapCompletable`

maskaravivek replied:
Just updated the PR.

macgills commented on app/src/main/java/fr/free/nrw/commons/contributions/ContributionDao.java:
```diff
@@ -20,20 +22,36 @@
 
   @Insert(onConflict = OnConflictStrategy.REPLACE)
-  public abstract void save(Contribution contribution);
+  public abstract void saveSyncronous(Contribution contribution);
```
`synchronous`

macgills commented:
@neslihanturan could you give this a quick test? I think you have a far greater number of contributions than I do.

Otherwise looks good, has my approval

neslihanturan commented:
Scrolling is a lot smoother, my first insights about this Pr is very positive. Can you emphasize if there is a function that must be tested.

macgills commented:
@neslihanturan I just wanted verification a large data set is performant, that is good enough for my curiosity
